### PR TITLE
Arbitrary size concat elements utf8

### LINKED ISSUE #1748: `concat_elements_utf8` should accept arbitrary number of input arrays.
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
This is a follow-up of #1720.
We should support concatenating arbitrary number of arrays.
As rust does not support variadic function so far, we can refer to the `concat` function: https://github.com/apache/arrow-rs/blob/master/arrow/src/compute/kernels/concat.rs#L53

### PR DESCRIPTION
**Pull request by Ismail-Maj**
# Which issue does this PR close?

Closes #1748.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

Should be merged after #1780.

### CODE REVIEW

File: arrow/src/compute/kernels/concat_elements.rs [hunk 1]
```diff
@@ -129,21 +151,50 @@ mod tests {
...
+        );
+    }
+
+    #[test]
+    fn test_string_concat_multiple() {
+        let foo = &StringArray::from(vec![Some("f"), Some("o"), Some("o"), None]);
+        let bar = &StringArray::from(vec![None, Some("b"), Some("a"), Some("r")]);
+        let baz = &StringArray::from(vec![Some("b"), None, Some("a"), Some("z")]);
+
+        let output = concat_elements_utf8(&[&foo, &bar, &baz]).unwrap();
```
Comment (Ismail-Maj) [inline arrow/src/compute/kernels/concat_elements.rs:191]:
```suggestion
        let foo = StringArray::from(vec![Some("f"), Some("o"), Some("o"), None]);
        let bar = StringArray::from(vec![None, Some("b"), Some("a"), Some("r")]);
        let baz = StringArray::from(vec![Some("b"), None, Some("a"), Some("z")]);

        let output = concat_elements_utf8(&[&foo, &bar, &baz]).unwrap();
```

File: arrow/src/compute/kernels/concat_elements.rs [hunk 2]
```diff
@@ -29,54 +29,76 @@ use crate::error::{ArrowError, Result};
...
 ///
-/// An error will be returned if `left` and `right` have different lengths
+/// An error will be returned if the [`StringArray`] are of different lengths.
 pub fn concat_elements_utf8<Offset: OffsetSizeTrait>(
-    left: &GenericStringArray<Offset>,
-    right: &GenericStringArray<Offset>,
+    arrays: &[&GenericStringArray<Offset>],
 ) -> Result<GenericStringArray<Offset>> {
-    if left.len() != right.len() {
+    if arrays.len() < 2 {
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/concat_elements.rs:39]: We should follow the pattern in `concat` kernel: https://github.com/apache/arrow-rs/blob/master/arrow/src/compute/kernels/concat.rs#L53-L61

If you prefer FP style, you could write like this (not tested): 
```rust
match array {
    [] => Err(...),
    [arr] => Ok((arr.slice(0, arr.len())),
    [arr, ..] => {// your code here},
}
```

File: arrow/src/compute/kernels/concat_elements.rs [hunk 3]
```diff
@@ -29,54 +29,75 @@ use crate::error::{ArrowError, Result};
...
+        .collect::<Vec<_>>();
 
     let mut output_values = BufferBuilder::<u8>::new(
-        left_values.len() + right_values.len()
-            - left_offsets[0].to_usize().unwrap()
-            - right_offsets[0].to_usize().unwrap(),
+        data_values
+            .iter()
+            .zip(offsets.iter_mut())
+            .map(|(data, offset)| data.len() - offset.peek().unwrap().to_usize().unwrap())
```
Comment (Ismail-Maj) [inline arrow/src/compute/kernels/concat_elements.rs:81]: won't panic because offset is never empty after this gets merged #1825

Comment (tustvold) [reply arrow/src/compute/kernels/concat_elements.rs:81]: Yes, offsets can be empty. Should be a simple case of adding something like this at the top of the function

```
if arrays[0].is_empty() {
  return make_empty(arrays[0].data_type())
}
```

**Review by tustvold — COMMENTED** [review]
I would be very interested to see some benchmarks of this, this has introduced a number of additional memory allocations and a significantly more complicated loop body, it may not matter, but it would be good to verify either way...

**Comment by Ismail-Maj** [comment]
Thank you for the review @tustvold!
I'll work on a benchmark, I'm wondering if we could use [`smallvec`](https://github.com/servo/rust-smallvec) to avoid heap allocating with only a few arrays as input.

**Comment by alamb** [comment]
FYI I think @tustvold  is out this week

Another way to move this PR forward might be to make this a new kernel (rather than modifying the existing `concat_elements_utf8`), perhaps something like `concat_n_elements_utf8` or `concat_elements_utf8_many`?

That way we aren't worried about a potential regression in performance and people can optimize the new kernel in the future if that is desired

**Comment by HaoYang670** [comment]
> I'll work on a benchmark.

@Ismail-Maj  This could be done as a follow-up PR.

**Review by tustvold — APPROVED** [review]
Sorry for the delay, thank you :+1:

**Pull request merged**